<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Exploratory spatial data analysis with Voyager • VoyagerWorkshop</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Exploratory spatial data analysis with Voyager">
<meta property="og:description" content="VoyagerWorkshop">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">VoyagerWorkshop</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.99</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/workshop.html">Exploratory spatial data analysis with Voyager</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/lambdamoses/VoyagerWorkshop" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Exploratory spatial data analysis with
Voyager</h1>
                        <h4 data-toc-skip class="author">Lambda
Moses</h4>
                  <a class="author_email" href="mailto:#"></a><a href="mailto:dlu2@caltech.edu" class="email">dlu2@caltech.edu</a>
      
                              <h4 data-toc-skip class="author">Lior
Pachter</h4>
                  <a class="author_email" href="mailto:#"></a><a href="mailto:lpachter@caltech.edu" class="email">lpachter@caltech.edu</a>
      
                  
            <h4 data-toc-skip class="date">2023-07-19</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/lambdamoses/VoyagerWorkshop/blob/HEAD/vignettes/workshop.Rmd" class="external-link"><code>vignettes/workshop.Rmd</code></a></small>
      <div class="hidden name"><code>workshop.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Slides TBD. The following R packages are used in this workshop, which
are all on CRAN or Bioconductor. Bioconductor 3.17 is used in the
workshop presented at Bioc2023.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SpatialFeatureExperiment" class="external-link">SpatialFeatureExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/voyager" class="external-link">Voyager</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SFEData" class="external-link">SFEData</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/alexcb/rjson" class="external-link">rjson</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://Matrix.R-forge.R-project.org" class="external-link">Matrix</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-spatial.github.io/sf/" class="external-link">sf</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://scales.r-lib.org" class="external-link">scales</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel" class="external-link">BiocParallel</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/" class="external-link">tibble</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org" class="external-link">tidyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/thomasp85/scico" class="external-link">scico</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="spatialfeatureexperiment">
<code>SpatialFeatureExperiment</code><a class="anchor" aria-label="anchor" href="#spatialfeatureexperiment"></a>
</h3>
<p><code>SpatialFeatureExperiment</code> (SFE) is a new <a href="http://adv-r.had.co.nz/S4.html" class="external-link">S4</a> class built on top of <a href="https://bioconductor.org/packages/release/bioc/html/SpatialExperiment.html" class="external-link"><code>SpatialExperiment</code></a>
(SPE). <code>SpatialFeatureExperiment</code> incorporates geometries and
geometry operations with the <a href="https://cran.r-project.org/web/packages/sf/index.html" class="external-link"><code>sf</code></a>
package. Examples of supported geometries are Visium spots represented
with polygons corresponding to their size, cell or nuclei segmentation
polygons, tissue boundary polygons, pathologist annotation of
histological regions, and transcript spots of genes. Using
<code>sf</code>, <code>SpatialFeatureExperiment</code> leverages the
GEOS C++ libraries underlying <code>sf</code> for geometry operations,
including algorithms for for determining whether geometries intersect,
finding intersection geometries, buffering geometries with margins, etc.
A schematic of the SFE object is shown below:</p>
<div class="figure">
<img src="sfe_schematics.png" alt="SpatialFeatureExperiment expands on SpatialExperiment by adding column, row, and annotation geometries and spatial graphs. This is explained in detail in the following paragraphs." width="100%"><p class="caption">
Schematics of the SFE object
</p>
</div>
<p>Below is a list of SFE features that extend the SPE object:</p>
<ul>
<li>
<code>colGeometries</code> are <code>sf</code> data frames
associated with the entities that correspond to columns of the gene
count matrix, such as Visium spots or cells. The geometries in the
<code>sf</code> data frames can be Visium spot centroids, Visium spot
polygons, or for datasets with single cell resolution, cell or nuclei
segmentations. Multiple <code>colGeometries</code> can be stored in the
same SFE object, such as one for cell segmentation and another for
nuclei segmentation. There can be non-spatial, attribute columns in a
<code>colGeometry</code> rather than <code>colData</code>, because the
<code>sf</code> class allows users to specify how attributes relate to
geometries, such as “constant”, “aggregate”, and “identity”. See the
<code>agr</code> argument of the <a href="https://r-spatial.github.io/sf/reference/sf.html" class="external-link"><code>st_sf</code>
documentation</a>.</li>
<li>
<code>colGraphs</code> are spatial neighborhood graphs of cells or
spots. The graphs have class <code>listw</code> (<code>spdep</code>
package), and the <code>colPairs</code> of
<code>SingleCellExperiment</code> was not used so no conversion is
necessary to use the numerous spatial dependency functions from
<code>spdep</code>, such as those for Moran’s I, Geary’s C, Getis-Ord
Gi*, LOSH, etc. Conversion is also not needed for other classical
spatial statistics packages such as <code>spatialreg</code> and
<code>adespatial</code>.</li>
<li>
<code>rowGeometries</code> are similar to
<code>colGeometries</code>, but support entities that correspond to rows
of the gene count matrix, such as genes. A potential use case is to
store transcript spots for each gene in smFISH or in situ sequencing
based datasets.</li>
<li>
<code>rowGraphs</code> are similar to <code>colGraphs</code>. A
potential use case may be spatial colocalization of transcripts of
different genes.</li>
<li>
<code>annotGeometries</code> are <code>sf</code> data frames
associated with the dataset but not directly with the gene count matrix,
such as tissue boundaries, histological regions, cell or nuclei
segmentation in Visium datasets, and etc. These geometries are stored in
this object to facilitate plotting and using <code>sf</code> for
operations such as to find the number of nuclei in each Visium spot and
which histological regions each Visium spot intersects. Unlike
<code>colGeometries</code> and <code>rowGeometries</code>, the number of
rows in the <code>sf</code> data frames in <code>annotGeometries</code>
is not constrained by the dimension of the gene count matrix and can be
arbitrary.</li>
<li>
<code>annotGraphs</code> are similar to <code>colGraphs</code> and
<code>rowGraphs</code>, but are for entities not directly associated
with the gene count matrix, such as spatial neighborhood graphs for
nuclei in Visium datasets, or other objects like myofibers. These graphs
are relevant to <code>spdep</code> analyses of attributes of these
geometries such as spatial autocorrelation in morphological metrics of
myofibers and nuclei. With geometry operations with <code>sf</code>,
these attributes and results of analyses of these attributes
(e.g. spatial regions defined by the attributes) may be related back to
gene expression.</li>
<li>
<code>localResults</code> are similar to <a href="https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html#3_Adding_low-dimensional_representations" class="external-link"><code>reducedDims</code>
in <code>SingleCellExperiment</code></a>, but stores results from
univariate and bivariate local spatial analysis results, such as from <a href="https://r-spatial.github.io/spdep/reference/localmoran.html" class="external-link"><code>localmoran</code></a>,
<a href="https://r-spatial.github.io/spdep/reference/localG.html" class="external-link">Getis-Ord
Gi*</a>, and <a href="https://r-spatial.github.io/spdep/reference/LOSH.html" class="external-link">local
spatial heteroscedasticity (LOSH)</a>. Unlike in
<code>reducedDims</code>, for each type of results (type is the type of
analysis such as Getis-Ord Gi*), each feature (e.g. gene) or pair of
features for which the analysis is performed has its own results. The
local spatial analyses can also be performed for attributes of
<code>colGeometries</code> and <code>annotGeometries</code> in addition
to gene expression and <code>colData</code>. Results of multivariate
spatial analysis such as <a href="https://cran.r-project.org/web/packages/adespatial/vignettes/tutorial.html#multispati-analysis" class="external-link">MULTISPATI
PCA</a> can be stored in <code>reducedDims</code>.</li>
</ul>
</div>
<div class="section level3">
<h3 id="voyager">
<code>Voyager</code><a class="anchor" aria-label="anchor" href="#voyager"></a>
</h3>
<div class="figure">
<img src="voyager.png" alt="Schematics of the Voyager object" width="100%"><p class="caption">
Schematics of the Voyager object
</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="part-1-the-sfe-class">Part 1: The SFE class<a class="anchor" aria-label="anchor" href="#part-1-the-sfe-class"></a>
</h2>
<div class="section level3">
<h3 id="create-an-sfe-object">Create an SFE object<a class="anchor" aria-label="anchor" href="#create-an-sfe-object"></a>
</h3>
<div class="section level4">
<h4 id="visium-space-ranger-output">Visium Space Ranger output<a class="anchor" aria-label="anchor" href="#visium-space-ranger-output"></a>
</h4>
<p>10x Genomics Space Ranger output from a Visium experiment can be read
in a similar manner as in <code>SpatialExperiment</code>; the
<code>SpatialFeatureExperiment</code> SFE object has the
<code>spotPoly</code> column geometry for the spot polygons. If the
filtered matrix (i.e. only spots in the tissue) is read in, then a
column graph called <code>visium</code> will also be present for the
spatial neighborhood graph of the Visium spots on the tissue. The graph
is not computed if all spots are read in regardless of whether they are
on tissue.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Example from SpatialExperiment</span></span>
<span><span class="va">dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"10xVisium"</span><span class="op">)</span>, </span>
<span>  package <span class="op">=</span> <span class="st">"SpatialExperiment"</span><span class="op">)</span></span>
<span>  </span>
<span><span class="va">sample_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"section1"</span>, <span class="st">"section2"</span><span class="op">)</span></span>
<span><span class="op">(</span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir</span>, <span class="va">sample_ids</span>, <span class="st">"outs"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "/usr/local/lib/R/site-library/SpatialExperiment/extdata/10xVisium/section1/outs"</span></span>
<span><span class="co">#&gt; [2] "/usr/local/lib/R/site-library/SpatialExperiment/extdata/10xVisium/section2/outs"</span></span></code></pre></div>
<p>The results for each tissue capture should be in the
<code>outs</code> directory. Inside the <code>outs</code> directory
there are two directories: <code>raw_reature_bc_matrix</code> has the
unfiltered gene count matrix, and <code>spatial</code> has the spatial
information.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">samples</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "raw_feature_bc_matrix" "spatial"</span></span></code></pre></div>
<p>The <a href="https://bioconductor.org/packages/release/bioc/html/DropletUtils.html" class="external-link"><code>DropletUtils</code></a>
package has a function <code>read10xCounts()</code> which reads the gene
count matrix. SPE reads in the spatial information, and SFE uses the
spatial information to construct Visium spot polygons and spatial
neighborhood graphs. Inside the <code>spatial</code> directory:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">samples</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"spatial"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "scalefactors_json.json"    "tissue_lowres_image.png"  </span></span>
<span><span class="co">#&gt; [3] "tissue_positions_list.csv"</span></span></code></pre></div>
<p><code>tissue_lowres_image.png</code> is a low resolution image of the
tissue.</p>
<p>Inside the <code>scalefactors_json.json</code> file:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/rjson/man/fromJSON.html" class="external-link">fromJSON</a></span><span class="op">(</span>file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">samples</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"spatial"</span>, <span class="st">"scalefactors_json.json"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; $spot_diameter_fullres</span></span>
<span><span class="co">#&gt; [1] 89.44476</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $tissue_hires_scalef</span></span>
<span><span class="co">#&gt; [1] 0.1701114</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $fiducial_diameter_fullres</span></span>
<span><span class="co">#&gt; [1] 144.4877</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $tissue_lowres_scalef</span></span>
<span><span class="co">#&gt; [1] 0.05103343</span></span></code></pre></div>
<p><code>spot_diameter_fullres</code> is the diameter of each Visium
spot in the full resolution H&amp;E image in pixels.
<code>tissue_hires_scalef</code> and <code>tissue_lowres_scalef</code>
are the ratio of the size of the high resolution (but not full
resolution) and low resolution H&amp;E image to the full resolution
image. <code>fiducial_diameter_fullres</code> is the diameter of each
fiducial spot used to align the spots to the H&amp;E image in pixels in
the full resolution image.</p>
<p>The <code>tissue_positions_list.csv</code> file contains information
for the spatial coordinates of the spots and whether each spot is in
tissue as automatically detected by Space Ranger or manually annotated
in the Loupe browser. If the polygon of the tissue boundary is
available, whether from image processing or manual annotation, geometric
operations as supported by the SFE package, which is based on the
<code>sf</code> package, can be used to find which spots intersect with
the tissue and which spots are contained in the tissue. Geometric
operations can also find the polygons of the intersections between spots
and the tissue, but the results can get messy since the intersections
can have not only polygons but also points and lines.</p>
<p>Now we read in the toy data that is in the Space Ranger output
format. Since Bioconductor version 3.17 (Voyager version 1.2.0), the
image is read as a <code>SpatRaster</code> object with the <a href="https://rspatial.github.io/terra/index.html" class="external-link"><code>terra</code></a>
package, so it is not loaded into memory unless necessary. When plotting
a large image, it will be downsampled and thus not fully loaded into
memory. The unit can be set with the <code>unit</code> argument, and can
be either pixels in full resolution image or microns. The latter is
calculated from the former based on spacing between spots, which is
known to be 100 microns.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/read10xVisiumSFE.html" class="external-link">read10xVisiumSFE</a></span><span class="op">(</span><span class="va">samples</span>, dirs <span class="op">=</span> <span class="va">samples</span>, sample_id <span class="op">=</span> <span class="va">sample_ids</span>, </span>
<span>                          type <span class="op">=</span> <span class="st">"sparse"</span>, data <span class="op">=</span> <span class="st">"raw"</span>, images <span class="op">=</span> <span class="st">"lowres"</span>, </span>
<span>                          unit <span class="op">=</span> <span class="st">"full_res_image_pixel"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 50 99 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(50): ENSMUSG00000051951 ENSMUSG00000089699 ...</span></span>
<span><span class="co">#&gt;   ENSMUSG00000005886 ENSMUSG00000101476</span></span>
<span><span class="co">#&gt; rowData names(1): symbol</span></span>
<span><span class="co">#&gt; colnames(99): AAACAACGAATAGTTC-1 AAACAAGTATCTCCCA-1 ...</span></span>
<span><span class="co">#&gt;   AAAGTCGACCCTCAGT-1-1 AAAGTGCCATCAATTA-1-1</span></span>
<span><span class="co">#&gt; colData names(4): in_tissue array_row array_col sample_id</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : pxl_col_in_fullres pxl_row_in_fullres</span></span>
<span><span class="co">#&gt; imgData names(4): sample_id image_id data scaleFactor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: full_res_image_pixel</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: spotPoly (POLYGON) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; section1: </span></span>
<span><span class="co">#&gt; section2:</span></span></code></pre></div>
<p>Space Ranger output includes the gene count matrix, spot coordinates,
and spot diameter. The Space Ranger output does NOT include nuclei
segmentation or pathologist annotation of histological regions. Extra
image processing, such as with ImageJ and QuPath, are required for those
geometries.</p>
</div>
<div class="section level4">
<h4 id="from-scratch">From scratch<a class="anchor" aria-label="anchor" href="#from-scratch"></a>
</h4>
<p>An SFE object can be constructed from scratch with the assay matrices
and metadata. In this toy example, <code>dgCMatrix</code> is used, but
since SFE inherits from SingleCellExperiment (SCE), other types of
arrays supported by SCE such as delayed arrays should also work.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visium barcode location from Space Ranger</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"visium_row_col"</span><span class="op">)</span></span>
<span><span class="va">coords1</span> <span class="op">&lt;-</span> <span class="va">visium_row_col</span><span class="op">[</span><span class="va">visium_row_col</span><span class="op">$</span><span class="va">col</span> <span class="op">&lt;</span> <span class="fl">6</span> <span class="op">&amp;</span> <span class="va">visium_row_col</span><span class="op">$</span><span class="va">row</span> <span class="op">&lt;</span> <span class="fl">6</span>,<span class="op">]</span></span>
<span><span class="va">coords1</span><span class="op">$</span><span class="va">row</span> <span class="op">&lt;-</span> <span class="va">coords1</span><span class="op">$</span><span class="va">row</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Random toy sparse matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">29</span><span class="op">)</span></span>
<span><span class="va">col_inds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">13</span>, <span class="fl">13</span><span class="op">)</span></span>
<span><span class="va">row_inds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">13</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">values</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">13</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/sparseMatrix.html" class="external-link">sparseMatrix</a></span><span class="op">(</span>i <span class="op">=</span> <span class="va">row_inds</span>, j <span class="op">=</span> <span class="va">col_inds</span>, x <span class="op">=</span> <span class="va">values</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">coords1</span><span class="op">$</span><span class="va">barcode</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">LETTERS</span>, <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p>This should be sufficient to create an SPE object, and an SFE object,
even though no <code>sf</code> data frame was constructed for the
geometries. The constructor behaves similarly to the SPE constructor.
The centroid coordinates of the Visium spots in the example can be
converted into spot polygons with the <code>spotDiameter</code>
argument, which can also be relevant to other technologies with round
spots or beads, such as Slide-seq. Spot diameter in pixels in full
resolution images can be found in the
<code>scalefactors_json.json</code> file in Space Ranger output.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/SpatialFeatureExperiment.html" class="external-link">SpatialFeatureExperiment</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">mat</span><span class="op">)</span>, colData <span class="op">=</span> <span class="va">coords1</span>,</span>
<span>                                spatialCoordsNames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"col"</span>, <span class="st">"row"</span><span class="op">)</span>,</span>
<span>                                spotDiameter <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span></code></pre></div>
<p>More geometries and spatial graphs can be added after calling the
constructor.</p>
<p>Geometries can also be supplied in the constructor.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Convert regular data frame with coordinates to sf data frame</span></span>
<span><span class="va">cg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/df2sf.html" class="external-link">df2sf</a></span><span class="op">(</span><span class="va">coords1</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"col"</span>, <span class="st">"row"</span><span class="op">)</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"col"</span>, <span class="st">"row"</span><span class="op">)</span>, spotDiameter <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">cg</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span></span>
<span><span class="va">sfe3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/SpatialFeatureExperiment.html" class="external-link">SpatialFeatureExperiment</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">mat</span><span class="op">)</span>, colGeometries <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>foo <span class="op">=</span> <span class="va">cg</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="part-2-voyager-esda">Part 2: Voyager ESDA<a class="anchor" aria-label="anchor" href="#part-2-voyager-esda"></a>
</h2>
<div class="section level3">
<h3 id="dataset">Dataset<a class="anchor" aria-label="anchor" href="#dataset"></a>
</h3>
<p>The dataset used in this vignette is from the paper <a href="https://doi.org/10.1038/s42003-021-02810-x" class="external-link">Large-scale
integration of single-cell transcriptomic data captures transitional
progenitor states in mouse skeletal muscle regeneration</a> <span class="citation">(McKellar et al. 2021)</span>. Notexin was injected
into the tibialis anterior muscle of mice to induce injury, and the
healing muscle was collected 2, 5, and 7 days post injury for Visium
analysis. The dataset in this vignette is from the timepoint at day 2.
The vignette starts with a <code>SpatialFeatureExperiment</code> (SFE)
object.</p>
<p>The gene count matrix was directly downloaded <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM4904759" class="external-link">from
GEO</a>. All 4992 spots, whether in tissue or not, are included. The
H&amp;E image was used for nuclei and myofiber segmentation. A subset of
nuclei from randomly selected regions from all 3 timepoints were
manually annotated to train a StarDist model to segment the rest of the
nuclei, and the myofibers were all manually segmented. The tissue
boundary was found by thresholding in OpenCV, and small polygons were
removed as they are likely to be debris. Spot polygons were constructed
with the spot centroid coordinates and diameter in the Space Ranger
output. The <code>in_tissue</code> column in <code>colData</code>
indicates which spot polygons intersect the tissue polygons, and is
based on <code><a href="https://r-spatial.github.io/sf/reference/geos_binary_pred.html" class="external-link">st_intersects()</a></code>.</p>
<p>Tissue boundary, nuclei, myofiber, and Visium spot polygons are
stored as <code>sf</code> data frames in the SFE object. See <a href="https://bioconductor.org/packages/devel/bioc/vignettes/SpatialFeatureExperiment/inst/doc/SFE.html" class="external-link">the
vignette of <code>SpatialFeatureExperiment</code></a> for more details
on the structure of the SFE object. The SFE object of this dataset is
provided in the <code>SFEData</code> package; we begin by downloading
the data and loading it into R.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/McKellarMuscleData.html" class="external-link">McKellarMuscleData</a></span><span class="op">(</span><span class="st">"full"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; see ?SFEData and browseVignettes('SFEData') for documentation</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 15123 4992 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(15123): ENSMUSG00000025902 ENSMUSG00000096126 ...</span></span>
<span><span class="co">#&gt;   ENSMUSG00000064368 ENSMUSG00000064370</span></span>
<span><span class="co">#&gt; rowData names(6): Ensembl symbol ... vars cv2</span></span>
<span><span class="co">#&gt; colnames(4992): AAACAACGAATAGTTC AAACAAGTATCTCCCA ... TTGTTTGTATTACACG</span></span>
<span><span class="co">#&gt;   TTGTTTGTGTAAATTC</span></span>
<span><span class="co">#&gt; colData names(12): barcode col ... prop_mito in_tissue</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : imageX imageY</span></span>
<span><span class="co">#&gt; imgData names(1): sample_id</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: full_res_image_pixels</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: spotPoly (POLYGON) </span></span>
<span><span class="co">#&gt; annotGeometries: tissueBoundary (POLYGON), myofiber_full (POLYGON), myofiber_simplified (POLYGON), nuclei (POLYGON), nuclei_centroid (POINT) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; Vis5A:</span></span></code></pre></div>
<p>The H&amp;E image of this section: <img src="https://raw.githubusercontent.com/pachterlab/voyager/documentation/vignettes/tissue_lowres_5a.jpeg" alt="A cross section of mouse muscle is slightly off center to the lower left. In the middle of the tissue is the notexin injury site with leukocyte infiltration and fewer myofibers. The rest of the tissue section is tightly packed with myofibers."></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="st">"tissue_lowres_5a.jpeg"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://raw.githubusercontent.com/pachterlab/voyager/main/vignettes/tissue_lowres_5a.jpeg"</span>,</span>
<span>                  destfile <span class="op">=</span> <span class="st">"tissue_lowres_5a.jpeg"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>The image can be added to the SFE object and plotted behind the
geometries, and needs to be flipped to align to the spots because the
origin is at the top left for the image but bottom left for
geometries.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu">addImg</span><span class="op">(</span><span class="va">sfe</span>, file <span class="op">=</span> <span class="st">"tissue_lowres_5a.jpeg"</span>, sample_id <span class="op">=</span> <span class="st">"Vis5A"</span>, </span>
<span>              image_id <span class="op">=</span> <span class="st">"lowres"</span>, </span>
<span>              scale_fct <span class="op">=</span> <span class="fl">1024</span><span class="op">/</span><span class="fl">22208</span><span class="op">)</span></span>
<span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu">mirrorImg</span><span class="op">(</span><span class="va">sfe</span>, sample_id <span class="op">=</span> <span class="st">"Vis5A"</span>, image_id <span class="op">=</span> <span class="st">"lowres"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="exploratory-data-analysis">Exploratory data analysis<a class="anchor" aria-label="anchor" href="#exploratory-data-analysis"></a>
</h3>
<div class="section level4">
<h4 id="spots-in-tissue">Spots in tissue<a class="anchor" aria-label="anchor" href="#spots-in-tissue"></a>
</h4>
<p>While the example dataset has all Visium spots whether on tissue or
not, only spots that intersect tissue are used for further analyses.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "barcode"   "col"       "row"       "x"         "y"         "dia"      </span></span>
<span><span class="co">#&gt;  [7] "tissue"    "sample_id" "nCounts"   "nGenes"    "prop_mito" "in_tissue"</span></span></code></pre></div>
<p>Total UMI counts (<code>nCounts</code>), number of genes detected per
spot (<code>nGenes</code>), and the proportion of mitochondrially
encoded counts (<code>prop_mito</code>) have been precomputed and are in
<code>colData(sfe)</code>. The <code>plotSpatialFeature</code> function
can be used to visualize various attributes in space: the expression of
any gene, <code>colData</code> values, and geometry attributes in
<code>colGeometry</code> and <code>annotGeometry</code>. The Visium
spots are plotted as polygons reflecting their actual size relative to
the tissue, rather than as points, as is the case in other packages that
plot Visium data. The plotting of geometries is being performed under
the hood with <code>geom_sf</code>.</p>
<p>The tissue boundary was found by thresholding the H&amp;E image and
removing small polygons that are most likely debris. The
<code>in_tissue</code> column of <code>colData(sfe)</code> indicates
which Visium spot polygon intersects the tissue polygon; this can be
found with <code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotPred.html" class="external-link">SpatialFeatureExperiment::annotPred()</a></code>.</p>
<p>We demonstrate the use of <code>scran</code> <span class="citation">(Lun, Bach, and Marioni 2016)</span> for normalization
below, although we note that it is not necessarily the best approach to
normalizing spatial transcriptomics data. The problem of when and how to
normalize spatial transcriptomics data is non-trivial because, as the
<code>nCounts</code> plot in space shows above, spatial autocorrelation
is evident. Furthemrore, in Visium, reverse transcription occurs in situ
on the spots, but PCR amplification occurs after the cDNA is dissociated
from the spots. Artifacts may be subsequently introduced from the
amplification step, and these would not be associated with spatial
origin. Spatial artifacts may arise from the diffusion of transcripts
and tissue permeablization. However, given how the total counts seem to
correspond to histological regions, the total counts may have a
biological component and hence should not be treated as a technical
artifact to be normalized away as in scRNA-seq data normalization
methods. In other words, the issue of normalization for spatial
transcriptomics data, and Visium in particular, is complex and is
currently unsolved.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">in_tissue</span><span class="op">]</span></span>
<span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="va">sfe_tissue</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>,<span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#clusters &lt;- quickCluster(sfe_tissue)</span></span>
<span><span class="co">#sfe_tissue &lt;- computeSumFactors(sfe_tissue, clusters=clusters)</span></span>
<span><span class="co">#sfe_tissue &lt;- sfe_tissue[, sizeFactors(sfe_tissue) &gt; 0]</span></span>
<span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span></code></pre></div>
<p>Myofiber and nuclei segmentation polygons are available in this
dataset in the <code>annotGeometries</code> field. Myofibers were
manually segmented, and nuclei were segmented with <a href="https://github.com/stardist/stardist" class="external-link"><code>StarDist</code></a>
trained with a manually segmented subset.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotGeometries.html" class="external-link">annotGeometryNames</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "tissueBoundary"      "myofiber_full"       "myofiber_simplified"</span></span>
<span><span class="co">#&gt; [4] "nuclei"              "nuclei_centroid"</span></span></code></pre></div>
<div class="section level5">
<h5 id="from-myofibers-and-nuclei-to-visium-spots">From myofibers and nuclei to Visium spots<a class="anchor" aria-label="anchor" href="#from-myofibers-and-nuclei-to-visium-spots"></a>
</h5>
<p>The <code><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature()</a></code> function can also be used to
plot attributes of geometries, i.e. the non-geometry columns in the
<code>sf</code> data frames in the <code>rowGeometries</code>,
<code>colGeometries</code>, or <code>annotGeometries</code> fields of
the SFE object. For <code>rowGeometries</code> and
<code>colGeometries</code>, such columns which are associated with the
<code>sf</code> data frames rather than <code>rowData</code> or
<code>colData</code>, are allowed because one can specify how these
columns associate with the geometries (see <a href="https://r-spatial.github.io/sf/reference/st_agr.html" class="external-link"><code>st_agr</code></a>
and <a href="https://r-spatial.github.io/sf/reference/sf.html#details-1" class="external-link">documentation
of <code>st_sf</code></a>). When an attribute of an
<code>annotGeometry</code> is plotted along side gene expression or
<code>colData</code> or <code>colGeometry</code> attribute, the
<code>annotGeometry</code> attribute is plotted with a different color
palette to distinguish it from the column associated values.</p>
<p>The myofiber polygons from <code>annotGeometries</code> can be
plotted as shown below, colored by cross section area as observed in the
tissue section. The <code>aes_use</code> argument is set to
<code>color</code> rather than <code>fill</code> (default for polygons)
to only plot the Visium spot outlines to make the myofiber polygons more
visible. The <code>fill</code> argument is set to <code>NA</code> to
make the Visium spots look hollow, and the <code>size</code> argument
controls the thickness of the outlines. The <code>annot_aes</code>
argument specifies which column in the <code>annotGeometry</code> to use
to specify the values of an aesthstic, just like <code>aes</code> in
<code>ggplot2</code> (<code>aes_string</code> to be precise, since
<code>tidyeval</code> is not used here). The <code>annot_fixed</code>
argument (not used here) can set the fixed size, alpha, color, and etc.
for the <code>annotGeometry</code>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, features <span class="op">=</span> <span class="st">"nCounts"</span>, </span>
<span>                   colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>,</span>
<span>                   annotGeometryName <span class="op">=</span> <span class="st">"myofiber_simplified"</span>, </span>
<span>                   aes_use <span class="op">=</span> <span class="st">"color"</span>, linewidth <span class="op">=</span> <span class="fl">0.5</span>, fill <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>                   annot_aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"area"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-20-1.png" alt="Plot of Visium spots in tissue and myofiber polygons in physical space. Visium spots are colored by nCounts, and myofibers are colored by area." width="700"></p>
<p>The larger myofibers seem to have fewer total counts, possibly
because the larger size of these myofibers dilutes the transcripts. This
hints at the need for a normalization procedure.</p>
<p>With <code>SpatialFeatureExperiment</code>, we can find the number of
myofibers and nuclei that intersect each Visium spot. The predicate can
be <a href="https://r-spatial.github.io/sf/reference/geos_binary_pred.html" class="external-link">anything
implemented in <code>sf</code></a>, so for example, the number of nuclei
fully covered by each Visium spot can also be found. The default
predicate is <code><a href="https://r-spatial.github.io/sf/reference/geos_binary_pred.html" class="external-link">st_intersects()</a></code>.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">$</span><span class="va">n_myofibers</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotPred.html" class="external-link">annotNPred</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>,</span>
<span>             annotGeometryName <span class="op">=</span> <span class="st">"myofiber_simplified"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, features <span class="op">=</span> <span class="st">"n_myofibers"</span>, </span>
<span>                   colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>, image <span class="op">=</span> <span class="st">"lowres"</span>, color <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>                   linewidth <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-22-1.png" alt="Plot of Visium spots in tissue in physical space, colored by number of myofibers intersecting each spot." width="700"></p>
<p>There is no one-to-one mapping between Visium spots and myofibers.
However, we can relate attributes of myofibers to gene expression
detected at the Visium spots. One way to do so is to summarize the
attributes of all myofibers that intersect (or choose another better
predicate implemented in <code>sf</code>) each spot, such as to
calculate the mean, median, or sum. This can be done with the
<code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotSummary.html" class="external-link">annotSummary()</a></code> function in
<code>SpatialFeatureExperiment</code>. The default predicate is
<code><a href="https://r-spatial.github.io/sf/reference/geos_binary_pred.html" class="external-link">st_intersects()</a></code>, and the default summary function is
<code><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean()</a></code>.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">$</span><span class="va">mean_myofiber_area</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotSummary.html" class="external-link">annotSummary</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"spotPoly"</span>, <span class="st">"myofiber_simplified"</span>, </span>
<span>               annotColNames <span class="op">=</span> <span class="st">"area"</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span> <span class="co"># it always returns a data frame</span></span></code></pre></div>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># The gray spots don't intersect any myofiber</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"mean_myofiber_area"</span>, <span class="st">"spotPoly"</span>, image <span class="op">=</span> <span class="st">"lowres"</span>, </span>
<span>                   color <span class="op">=</span> <span class="st">"black"</span>, linewidth <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-24-1.png" alt="Plot of Visium spots in tissue in physical space, colored by the average area of myofibers that intersect each spot. The average area is higher near the mid-top right part of the tissue." width="700"></p>
<p>This reveals the relationship between the mean area of myofibers
intersecting each Visium spot and other aspects of the spots, such as
total counts and gene expression.</p>
<p>The NAs designate spots not intersecting any myofibers, e.g. those in
the inflammatory region.</p>
<p>In the <a href="https://pachterlab.github.io/Voyager/articles/vig2_visium_basic.html" class="external-link">Basic
Visium vignette</a>, we encountered two mysterious branches and two
clusters in the nGenes vs. nCounts plot and the proportion of
mitochondrial counts vs. nCounts plot. Now we see that the two clusters
seem to be related to myofiber size.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, x <span class="op">=</span> <span class="st">"nCounts"</span>, y <span class="op">=</span> <span class="st">"nGenes"</span>, colour_by <span class="op">=</span> <span class="st">"mean_myofiber_area"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-25-1.png" width="700"></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, x <span class="op">=</span> <span class="st">"nCounts"</span>, y <span class="op">=</span> <span class="st">"prop_mito"</span>, colour_by <span class="op">=</span> <span class="st">"mean_myofiber_area"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-26-1.png" width="700"></p>
</div>
<div class="section level5">
<h5 id="myofiber-types">Myofiber types<a class="anchor" aria-label="anchor" href="#myofiber-types"></a>
</h5>
<p>Marker genes: Myh7 (Type I, slow twitch, aerobic), Myh2 (Type IIa,
fast twitch, somewhat aerobic), Myh4 (Type IIb, fast twitch, anareobic),
Myh1 (Type IIx, fast twitch, anaerobic), from <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5526362/" class="external-link">this
protocol</a> <span class="citation">(Wang, Yue, and Kuang
2017)</span></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>I <span class="op">=</span> <span class="st">"Myh7"</span>, IIa <span class="op">=</span> <span class="st">"Myh2"</span>, IIb <span class="op">=</span> <span class="st">"Myh4"</span>, IIx <span class="op">=</span> <span class="st">"Myh1"</span><span class="op">)</span></span></code></pre></div>
<p>We first examine the Type I myofibers. This is a fast twitch muscle,
so we don’t expect many slow twitch Type I myofibers. Row names in
<code>sfe_tissue</code> are Ensembl IDs in order to avoid ambiguity as
sometimes multiple Ensembl IDs have the same gene symbol and some genes
have aliases. However, gene symbols are shorter and more human readable
than Ensembl IDs, and are better suited to display on plots. In the
<code><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature()</a></code> function and other functions in
<code>Voyager</code>, even when the row names are recorded as Ensembl
IDs, the <code>features</code> argument can take gene symbols if there
is a column called “symbols” in <code>rowData(sfe)</code>, where the
function converts the gene symbols to Ensembl IDs. By default, gene
symbols are shown on the plot, but the <code>show_symbol</code> argument
can be set to <code>FALSE</code> to show Ensembl IDs instead. If one
gene symbol matches multiple Ensembl IDs in the dataset, then a warning
will be given.</p>
<p>The <code>exprs_values</code> argument specifies the assay to use,
which is by default “logcounts”, i.e. the log normalized data. This
default may or may not be suitable in practice given that total UMI
counts may have biological relevance in spatial data. Therefore, we plot
both the raw counts and the log normalized counts:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Function specific for this vignette, with some hard coded values</span></span>
<span><span class="va">plot_counts_logcounts</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">sfe</span>, <span class="va">feature</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">feature</span>, <span class="st">"spotPoly"</span>,</span>
<span>                   annotGeometryName <span class="op">=</span> <span class="st">"myofiber_simplified"</span>, </span>
<span>                   annot_aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"area"</span><span class="op">)</span>, swap_rownames <span class="op">=</span> <span class="st">"symbol"</span>, </span>
<span>                   exprs_values <span class="op">=</span> <span class="st">"counts"</span>, aes_use <span class="op">=</span> <span class="st">"color"</span>, linewidth <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>                   fill <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Raw counts"</span><span class="op">)</span></span>
<span>  <span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">feature</span>, <span class="st">"spotPoly"</span>,</span>
<span>                   annotGeometryName <span class="op">=</span> <span class="st">"myofiber_simplified"</span>, </span>
<span>                   annot_aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"area"</span><span class="op">)</span>, swap_rownames <span class="op">=</span> <span class="st">"symbol"</span>, </span>
<span>                   exprs_values <span class="op">=</span> <span class="st">"logcounts"</span>, aes_use <span class="op">=</span> <span class="st">"color"</span>, linewidth <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>                   fill <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Log normalized counts"</span><span class="op">)</span></span>
<span>  <span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_annotation.html" class="external-link">plot_annotation</a></span><span class="op">(</span>title <span class="op">=</span> <span class="va">feature</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_counts_logcounts</span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="va">markers</span><span class="op">[</span><span class="st">"I"</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-29-1.png" alt="Raw and log normalized counts of Myh7, marker gene of type I myofiber, plotted side by side on Visium spots in space, with myofiber polygons colored by myofiber cross section area plotted in the background. Visium spots expressing Myh7 concentrate in the lower left part of the tissue where the myofibers tend to be smaller." width="700"></p>
<p>A marker gene for type IIa myofibers is shown above. It is
straightforward to modify the plotting to display markers for type IIb
and type IIx myofibers:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_counts_logcounts</span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="va">markers</span><span class="op">[</span><span class="st">"IIa"</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-30-1.png" alt="Raw and log normalized counts of Myh2, a marker gene of type IIa myofiber, plotted side by side on Visium spots in space, with myofiber polygons colored by myofiber cross section area plotted in the background. Visium spots expressing Myh2 concentrate in the lower left and upper left parts of the tissue where the myofibers tend to be smaller. Log normalized counts show a wider region with higher expression." width="700"></p>
<p>Type IIa myofibers also tend to be clustered together on left side of
the tissue.</p>
<p>As SFE inherits from SCE, the non-spatial EDA plots from the
<code>scater</code> package can also be used:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, x <span class="op">=</span> <span class="st">"mean_myofiber_area"</span>, y <span class="op">=</span> <span class="st">"prop_mito"</span>, </span>
<span>            colour_by <span class="op">=</span> <span class="va">markers</span><span class="op">[</span><span class="st">"IIa"</span><span class="op">]</span>, by_exprs_values <span class="op">=</span> <span class="st">"logcounts"</span>, </span>
<span>            swap_rownames <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 36 rows containing missing values (`geom_point()`).</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-31-1.png" alt="Scatter plot of mean area of myofibers intersecting each Visium spot in the x axis and proportion of mitochondrially encoded counts per spot in the y axis, with points colored by expression of Myh2." width="700"></p>
<p>Plotting proportion of mitochondrial counts vs. mean myofiber area,
we see two clusters, one with higher proportion of mitochondrial counts
and smaller area, and another with lower proportion of mitochondrial
counts and on average slightly larger area. Type IIa myofibers tend to
have smaller area and a larger proportion of mitochondrial counts.</p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="spatial-neighborhood-graphs">Spatial neighborhood graphs<a class="anchor" aria-label="anchor" href="#spatial-neighborhood-graphs"></a>
</h3>
<p>A spatial neighborhood graph is required to compute spatial
dependency metrics such as Moran’s I and Geary’s C. The
<code>SpatialFeatureExperiment</code> package wraps methods in
<code>spdep</code> to find spatial neighborhood graphs, which are stored
within the SFE object (see <code>spdep</code> documentation for
<code>gabrielneigh()</code>, <code>knearneigh()</code>,
<code>poly2nb()</code>, and <code>tri2nb()</code>). The
<code>Voyager</code> package then uses these graphs for spatial
dependency analyses, again based on <code>spdep</code> in this first
version, but methods from other geospatial packages, some of which also
use the spatial neighborhood graphs, may be added later.</p>
<p>For Visium, where the spots are in a hexagonal grid, the spatial
neighborhood graph is straightforward. However, for spatial technologies
with single cell resolution, e.g. MERFISH, different methods can be used
to find the spatial neighborhood graph. In this example, the method
“poly2nb” was used for myofibers, and it identifies myofiber polygons
that physically touch each other. <code>zero.policy = TRUE</code> will
allow for singletons, i.e. nodes without neighbors in the graph; in the
inflamed region, there are more singletons. We have not yet benchmarked
spatial neighborhood construction methods to determine which is the
“best” for different technologies; the particular method used here is
for demonstration purposes and may not be the best in practice:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"visium"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/findVisiumGraph.html" class="external-link">findVisiumGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">annotGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"myofiber_poly2nb"</span><span class="op">)</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/findSpatialNeighbors-SpatialFeatureExperiment-method.html" class="external-link">findSpatialNeighbors</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, type <span class="op">=</span> <span class="st">"myofiber_simplified"</span>, MARGIN <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                       method <span class="op">=</span> <span class="st">"poly2nb"</span>, zero.policy <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>The <code><a href="https://rdrr.io/pkg/Voyager/man/plotColGraph.html" class="external-link">plotColGraph()</a></code> function plots the graph in space
associated with a <code>colGeometry</code>, along with the geometry of
interest.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotColGraph.html" class="external-link">plotColGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, colGraphName <span class="op">=</span> <span class="st">"visium"</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-33-1.png" alt="Spatial neighborhood graph of Visium spots that intersect tissue." width="700"></p>
<p>Similarly, the <code><a href="https://rdrr.io/pkg/Voyager/man/plotColGraph.html" class="external-link">plotAnnotGraph()</a></code> function plots the graph
associated with an <code>annotGeometry</code>, along with the geometry
of interest.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotColGraph.html" class="external-link">plotAnnotGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, annotGraphName <span class="op">=</span> <span class="st">"myofiber_poly2nb"</span>, </span>
<span>               annotGeometryName <span class="op">=</span> <span class="st">"myofiber_simplified"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-34-1.png" alt="Spatial neighborhood graph of myofibers, where each edge connects two myofibers that touch." width="700"></p>
<p>There is no <code>plotRowGraph</code> yet since we haven’t worked
with a dataset where spatial graphs related to genes are relevant,
although the SFE object supports row graphs.</p>
</div>
<div class="section level3">
<h3 id="exploratory-spatial-data-analysis">Exploratory <em>spatial</em> data analysis<a class="anchor" aria-label="anchor" href="#exploratory-spatial-data-analysis"></a>
</h3>
<p>All spatial autocorrelation metrics in this package can be computed
directly on a vector or a matrix rather than an SFE object. The user
interface emulates those of dimension reductions in the
<code>scater</code> package (e.g. <code><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">calculateUMAP()</a></code> that
takes in a matrix or SCE object and returns a matrix, and
<code><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP()</a></code> that takes in an SCE object and adds the results
to the <code>reducedDims</code> field of the SCE object). So
<code>calculate*</code> functions take in a matrix or an SFE object and
directly return the results (format of the results depends on the
structure of the results), while <code>run*</code> functions take in an
SFE object and add the results to the object. In addition,
<code>colData*</code> functions compute the metrics for numeric
variables in <code>colData</code>. <code>colGeometry*</code> functions
compute the metrics for numeric columns in a <code>colGeometry</code>.
<code>annotGeometry*</code> functions compute the metrics for numeric
columns in a <code>annotGeometry</code>.</p>
</div>
<div class="section level3">
<h3 id="univariate-global">Univariate global<a class="anchor" aria-label="anchor" href="#univariate-global"></a>
</h3>
<p><code>Voyager</code> supports many univariate global spatial
autocorrelation implemented in <code>spdep</code> for ESDA: Moran’s I
and Geary’s C, permutation testing for Moran’s I and Geary’s C, Moran
plot, and correlograms. In addition, beyond <code>spdep</code>,
<code>Voyager</code> can cluster Moran plots and correlograms. Plotting
functions taking in SFE objects are implemented to plot the results with
<code>ggplot2</code> and with more customization options than
<code>spdep</code> plotting functions. The functions
<code><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">calculateUnivariate()</a></code>, <code><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">runUnivariate()</a></code>,
<code><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">colDataUnivariate()</a></code>, <code><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">colGeometryUnivariate()</a></code>,
and <code><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">annotGeometryUnivariate()</a></code> compute univariate spatial
statistics. The argument <code>type</code>, which indicates the
corresponding function names in <code>spdep</code>, determines which
spatial statistics are computed.</p>
<p>All univariate global methods in <code>Voyager</code> are listed
here:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/listSFEMethods.html" class="external-link">listSFEMethods</a></span><span class="op">(</span>variate <span class="op">=</span> <span class="st">"uni"</span>, scope <span class="op">=</span> <span class="st">"global"</span><span class="op">)</span></span>
<span><span class="co">#&gt;              name                                           description</span></span>
<span><span class="co">#&gt; 1           moran                                             Moran's I</span></span>
<span><span class="co">#&gt; 2           geary                                             Geary's C</span></span>
<span><span class="co">#&gt; 3        moran.mc                    Moran's I with permutation testing</span></span>
<span><span class="co">#&gt; 4        geary.mc                    Geary's C with permutation testing</span></span>
<span><span class="co">#&gt; 5    sp.mantel.mc Mantel-Hubert spatial general cross product statistic</span></span>
<span><span class="co">#&gt; 6      moran.test                                        Moran's I test</span></span>
<span><span class="co">#&gt; 7      geary.test                                        Geary's C test</span></span>
<span><span class="co">#&gt; 8    globalG.test                                         Global G test</span></span>
<span><span class="co">#&gt; 9  sp.correlogram                                           Correlogram</span></span>
<span><span class="co">#&gt; 10      variogram                                  Variogram with model</span></span>
<span><span class="co">#&gt; 11  variogram_map                                         Variogram map</span></span></code></pre></div>
<p>When calling <code>calculate*variate()</code> or
<code>run*variate()</code>, the <code>type</code> (2nd) argument takes
either an <code>SFEMethod</code> object (see <code><a href="https://rdrr.io/pkg/Voyager/man/SFEMethod.html" class="external-link">SFEMethod()</a></code>
and <a href="https://pachterlab.github.io/voyager/articles/sfemethod.html" class="external-link">vignette
on <code>SFEMethod</code></a>) or a string that matches an entry in the
<code>name</code> column in the data frame returned by
<code><a href="https://rdrr.io/pkg/Voyager/man/listSFEMethods.html" class="external-link">listSFEMethods()</a></code>.</p>
<p>To demonstrate spatial autocorrelation in gene expression, top highly
variable genes (HVGs) are used. The HVGs are found with the
<code>scran</code> method.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span>
<span><span class="va">hvgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">dec</span>, n <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span></code></pre></div>
<p>A global statistic yields one result for the entire dataset.</p>
<div class="section level4">
<h4 id="morans-i">Moran’s I<a class="anchor" aria-label="anchor" href="#morans-i"></a>
</h4>
<p>There are several ways to quantify spatial autocorrelation, the most
common of which is Moran’s I:</p>
<p><span class="math display">\[
I = \frac{n}{\sum_{i=1}^n \sum_{j=1}^n w_{ij}} \frac{\sum_{i=1}^n
\sum_{j=1}^n w_{ij} (x_i - \bar{x})(x_j - \bar{x})}{\sum_{i=1}^n (x_i -
\bar{x})^2},
\]</span></p>
<p>where <span class="math inline">\(n\)</span> is the number of spots
or locations, <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span> are different locations, or spots in
the Visium context, <span class="math inline">\(x\)</span> is a variable
with values at each location, and <span class="math inline">\(w_{ij}\)</span> is a spatial weight, which can be
inversely proportional to distance between spots or an indicator of
whether two spots are neighbors, subject to various definitions of
neighborhood and whether to normalize the number of neighbors. The <a href="https://r-spatial.github.io/spdep/index.html" class="external-link"><code>spdep</code></a>
package uses the neighborhood.</p>
<p>Moran’s I can be understood as the Pearson correlation between the
value at each location and the average value at its neighbors. Just like
Pearson correlation, Moran’s I is generally bound between -1 and 1,
where positive value indicates positive spatial autocorrelation and
negative value indicates negative spatial autocorrelation.</p>
<p>Upon visual inspection, total UMI counts per spot seem to have
spatial autocorrelation. A spatial neighborhood graph is required to
compute Moran’s I, and is specified with the <code>listw</code>
argument.</p>
<p>For matrices, the rows are the features, as in the gene count
matrix.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Directly use vector or matrix, and multiple features can be specified at once</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">calculateUnivariate</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>, </span>
<span>                    type <span class="op">=</span> <span class="st">"moran"</span>,</span>
<span>                    listw <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"visium"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 2 rows and 2 columns</span></span>
<span><span class="co">#&gt;             moran         K</span></span>
<span><span class="co">#&gt;         &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; nCounts  0.528705   3.00082</span></span>
<span><span class="co">#&gt; nGenes   0.384028   3.88036</span></span></code></pre></div>
<p>“moran” is Moran’s I, and K is sample kurtosis.</p>
<p>To add the results to the SFE object, specifically for colData:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">colDataUnivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span><span class="op">)</span>,</span>
<span>                                colGraphName <span class="op">=</span> <span class="st">"visium"</span>, type <span class="op">=</span> <span class="st">"moran"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/colFeatureData.html" class="external-link">colFeatureData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="co">#&gt; DataFrame with 2 rows and 2 columns</span></span>
<span><span class="co">#&gt;         moran_Vis5A   K_Vis5A</span></span>
<span><span class="co">#&gt;           &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; nCounts    0.528705   3.00082</span></span>
<span><span class="co">#&gt; nGenes     0.384028   3.88036</span></span></code></pre></div>
<p>For <code>colData</code>, the results are added to
<code>colFeatureData(sfe)</code>, and features for which Moran’s I is
not calculated have NA. The column names of <code>featureData</code>
distinguishes between different samples (there’s only one sample in this
dataset), and are parsed by plotting functions.</p>
<p>To add the results to the SFE object, specifically for geometries:
Here “area” is the area of the cross section of each myofiber as seen in
this tissue section and “eccentricity” is the eccentricity of the
ellipse fitted to each myofiber.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Remember zero.policy = TRUE since there're singletons</span></span>
<span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">annotGeometryUnivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, type <span class="op">=</span> <span class="st">"moran"</span>,</span>
<span>                                      features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"area"</span>, <span class="st">"eccentricity"</span><span class="op">)</span>, </span>
<span>                                      annotGeometryName <span class="op">=</span> <span class="st">"myofiber_simplified"</span>,</span>
<span>                                      annotGraphName <span class="op">=</span> <span class="st">"myofiber_poly2nb"</span>, </span>
<span>                                      zero.policy <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotGeometries.html" class="external-link">annotGeometry</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"myofiber_simplified"</span><span class="op">)</span>, <span class="st">"featureData"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 6 rows and 2 columns</span></span>
<span><span class="co">#&gt;              moran_Vis5A   K_Vis5A</span></span>
<span><span class="co">#&gt;                &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; lyr.1                 NA        NA</span></span>
<span><span class="co">#&gt; area            0.327888   4.95675</span></span>
<span><span class="co">#&gt; perimeter             NA        NA</span></span>
<span><span class="co">#&gt; eccentricity    0.110938   3.26913</span></span>
<span><span class="co">#&gt; theta                 NA        NA</span></span>
<span><span class="co">#&gt; sine_theta            NA        NA</span></span></code></pre></div>
<p>For a non-geometry column in a <code>colGeometry</code>,
<code><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">colGeometryUnivariate()</a></code> is like
<code><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">annotGeometryUnivariate()</a></code> here, but none of the
<code>colGeometries</code> in this dataset has extra columns.</p>
<p>For gene expression, the <code>logcounts</code> assay is used by
default (use the <code>exprs_values</code> argument to change the
assay), though this may or may not be best practice. If the metrics are
computed for a large number of features, parallel computing is
supported, with <a href="https://bioconductor.org/packages/release/bioc/html/BiocParallel.html" class="external-link"><code>BiocParallel</code></a>,
with the <code>BPPARAM</code> argument.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">runUnivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, type <span class="op">=</span> <span class="st">"moran"</span>, features <span class="op">=</span> <span class="va">hvgs</span>, </span>
<span>                            colGraphName <span class="op">=</span> <span class="st">"visium"</span>, </span>
<span>                            BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">hvgs</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="co">#&gt; DataFrame with 6 rows and 8 columns</span></span>
<span><span class="co">#&gt;                               Ensembl      symbol            type     means</span></span>
<span><span class="co">#&gt;                           &lt;character&gt; &lt;character&gt;     &lt;character&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; ENSMUSG00000029304 ENSMUSG00000029304        Spp1 Gene Expression   1.63722</span></span>
<span><span class="co">#&gt; ENSMUSG00000050708 ENSMUSG00000050708        Ftl1 Gene Expression   2.37981</span></span>
<span><span class="co">#&gt; ENSMUSG00000050335 ENSMUSG00000050335      Lgals3 Gene Expression   1.43189</span></span>
<span><span class="co">#&gt; ENSMUSG00000021939 ENSMUSG00000021939        Ctsb Gene Expression   2.73117</span></span>
<span><span class="co">#&gt; ENSMUSG00000021190 ENSMUSG00000021190        Lgmn Gene Expression   1.11278</span></span>
<span><span class="co">#&gt; ENSMUSG00000018893 ENSMUSG00000018893          Mb Gene Expression   2.11118</span></span>
<span><span class="co">#&gt;                         vars       cv2 moran_Vis5A   K_Vis5A</span></span>
<span><span class="co">#&gt;                    &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; ENSMUSG00000029304   60.1583   22.4430    0.734937   1.63516</span></span>
<span><span class="co">#&gt; ENSMUSG00000050708  162.1931   28.6384    0.665563   1.81841</span></span>
<span><span class="co">#&gt; ENSMUSG00000050335   48.0739   23.4471    0.741474   1.68098</span></span>
<span><span class="co">#&gt; ENSMUSG00000021939  131.6232   17.6455    0.708362   1.86896</span></span>
<span><span class="co">#&gt; ENSMUSG00000021190   21.4505   17.3228    0.659916   1.66838</span></span>
<span><span class="co">#&gt; ENSMUSG00000018893   74.1782   16.6428    0.675840   1.82510</span></span></code></pre></div>
<p>Exercise: Compute Geary’s C on the highly variable genes.</p>
</div>
</div>
<div class="section level3">
<h3 id="univariate-local">Univariate local<a class="anchor" aria-label="anchor" href="#univariate-local"></a>
</h3>
<p>Local statistics yield a result at each location rather than the
whole dataset, while global statistics may obscure local heterogeneity.
See <span class="citation">(Fotheringham 2009)</span> for an interesting
discussion of relationships between global and local spatial statistics.
Local statistics are stored in the <code>localResults</code> field of
the SFE object, which can be accessed by the <code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/localResults.html" class="external-link">localResult()</a></code>
or <code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/localResults.html" class="external-link">localResults()</a></code> functions in the
<code>SpatialFeatureExperiment</code> package.</p>
<p>All univariate local methods in <code>Voyager</code> are listed
here:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/listSFEMethods.html" class="external-link">listSFEMethods</a></span><span class="op">(</span>variate <span class="op">=</span> <span class="st">"uni"</span>, scope <span class="op">=</span> <span class="st">"local"</span><span class="op">)</span></span>
<span><span class="co">#&gt;               name                                          description</span></span>
<span><span class="co">#&gt; 1       localmoran                                      Local Moran's I</span></span>
<span><span class="co">#&gt; 2  localmoran_perm                  Local Moran's I permutation testing</span></span>
<span><span class="co">#&gt; 3           localC                                      Local Geary's C</span></span>
<span><span class="co">#&gt; 4      localC_perm                  Local Geary's C permutation testing</span></span>
<span><span class="co">#&gt; 5           localG                                      Getis-Ord Gi(*)</span></span>
<span><span class="co">#&gt; 6      localG_perm             Getis-Ord Gi(*) with permutation testing</span></span>
<span><span class="co">#&gt; 7             LOSH                     Local spatial heteroscedasticity</span></span>
<span><span class="co">#&gt; 8          LOSH.mc Local spatial heteroscedasticity permutation testing</span></span>
<span><span class="co">#&gt; 9          LOSH.cs     Local spatial heteroscedasticity Chi-square test</span></span>
<span><span class="co">#&gt; 10      moran.plot                                   Moran scatter plot</span></span></code></pre></div>
<div class="section level4">
<h4 id="local-morans-i">Local Moran’s I<a class="anchor" aria-label="anchor" href="#local-morans-i"></a>
</h4>
<p>To recap, global Moran’s I is defined as</p>
<p><span class="math display">\[
I = \frac{n}{\sum_{i=1}^n \sum_{j=1}^n w_{ij}} \frac{\sum_{i=1}^n
\sum_{j=1}^n w_{ij} (x_i - \bar{x})(x_j - \bar{x})}{\sum_{i=1}^n (x_i -
\bar{x})^2}.
\]</span></p>
<p>Local Moran’s I <span class="citation">(Anselin 1995)</span> is
defined as</p>
<p><span class="math display">\[
I_i = (n-1)\frac{(x_i - \bar{x})\sum_{j=1}^n w_{ij} (x_j -
\bar{x})}{\sum_{i=1}^n (x_i - \bar{x})^2}.
\]</span></p>
<p>It’s similar to global Moran’s I, but the values at locations <span class="math inline">\(i\)</span> are not summed and there’s no
normalization by the sum of spatial weights.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">runUnivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, type <span class="op">=</span> <span class="st">"localmoran"</span>, features <span class="op">=</span> <span class="st">"Myh2"</span>,</span>
<span>                            colGraphName <span class="op">=</span> <span class="st">"visium"</span>, swap_rownames <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span></span></code></pre></div>
<p>It is useful to plot the log normalized Myh2 gene expression as
context to interpret the local results:</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, features <span class="op">=</span> <span class="st">"Myh2"</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>,</span>
<span>                   swap_rownames <span class="op">=</span> <span class="st">"symbol"</span>, image_id <span class="op">=</span> <span class="st">"lowres"</span>, color <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>                   linewidth <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-43-1.png" width="700"></p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotLocalResult.html" class="external-link">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"localmoran"</span>, features <span class="op">=</span> <span class="st">"Myh2"</span>, </span>
<span>                colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>,divergent <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                diverge_center <span class="op">=</span> <span class="fl">0</span>, image_id <span class="op">=</span> <span class="st">"lowres"</span>, </span>
<span>                swap_rownames <span class="op">=</span> <span class="st">"symbol"</span>, color <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>                linewidth <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-44-1.png" width="700"></p>
<p>We see that regions with higher Myh2 expression also have stronger
spatial autocorrelation. It is interesting to see how spatial
autocorrelation relates to gene expression level, much as finding how
variance relates to mean in the expression of each gene, which usually
indicates overdispersion compared to Poisson in scRNA-seq and Visium
data:</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>myh2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">$</span><span class="va">symbol</span> <span class="op">==</span> <span class="st">"Myh2"</span>,<span class="op">]</span>,</span>
<span>                 Ii <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/localResults.html" class="external-link">localResult</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"localmoran"</span>, <span class="st">"Myh2"</span>, </span>
<span>                                  swap_rownames <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span><span class="op">[</span>,<span class="st">"Ii"</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">myh2</span>, <span class="va">Ii</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Myh2 (log counts)"</span>, y <span class="op">=</span> <span class="st">"localmoran"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-45-1.png" width="700"></p>
<p>For this gene, Visium spots with higher expression also tend to have
higher local Moran’s I, but this may or may not apply to other
genes.</p>
<p>Local spatial analyses often return a matrix or data frame. The
<code><a href="https://rdrr.io/pkg/Voyager/man/plotLocalResult.html" class="external-link">plotLocalResult()</a></code> function has a default column for each
local spatial method, but other columns can be plotted as well. Use the
<code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/localResults.html" class="external-link">localResultAttrs()</a></code> function to see which columns are
present, and use the <code>attribute</code> argument to specify which
column to plot.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/localResults.html" class="external-link">localResultAttrs</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"localmoran"</span>, <span class="st">"Myh2"</span>, swap_rownames <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "Ii"             "E.Ii"           "Var.Ii"         "Z.Ii"          </span></span>
<span><span class="co">#&gt;  [5] "Pr(z != E(Ii))" "mean"           "median"         "pysal"         </span></span>
<span><span class="co">#&gt;  [9] "-log10p"        "-log10p_adj"</span></span></code></pre></div>
<p>Some local spatial methods return p-values at each location, in a
column with name like <code>Pr(z != E(Ii))</code>, where the test is two
sided (default, can be changed with the <code>alternative</code>
argument in <code><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">runUnivariate()</a></code> which is passed to the relevant
underlying function in <code>spdep</code>). Negative log of the p-value
is computed to facilitate visualization, and the p-value is corrected
for multiple hypothesis testing with <code>p.adjustSP()</code> in
<code>spdep</code>, where the number of tests is the number of neighbors
of each location rather than the total number of locations
(<code>-log10p_adj</code>).</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotLocalResult.html" class="external-link">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"localmoran"</span>, features <span class="op">=</span> <span class="st">"Myh2"</span>, </span>
<span>                colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>, attribute <span class="op">=</span> <span class="st">"-log10p_adj"</span>, divergent <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                diverge_center <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">)</span>, swap_rownames <span class="op">=</span> <span class="st">"symbol"</span>,</span>
<span>                image_id <span class="op">=</span> <span class="st">"lowres"</span>, color <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>                linewidth <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-47-1.png" width="700"></p>
<p>In this plot and all following plots of p-values, a divergent palette
is used to show locations that are significant after adjusting for
multiple testing and those that are not significant in different colors.
The center of the divergent palette is p = 0.05, so the bluish spots are
significant while a dark brown means <em>really</em> not
significant.</p>
<p>The “pysal” column displays the quadrants relative to the means in
the Moran plot. The result is similar to that from k-means clustering
shown above.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotLocalResult.html" class="external-link">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"localmoran"</span>, features <span class="op">=</span> <span class="st">"Myh2"</span>, </span>
<span>                colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>, attribute <span class="op">=</span> <span class="st">"pysal"</span>, </span>
<span>                swap_rownames <span class="op">=</span> <span class="st">"symbol"</span>, image_id <span class="op">=</span> <span class="st">"lowres"</span>, color <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>                linewidth <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-48-1.png" width="700"></p>
<p>Exercise: Compute local spatial heteroscedasticity (LOSH) on Myh2 and
plot the results.</p>
</div>
</div>
<div class="section level3">
<h3 id="bivariate">Bivariate<a class="anchor" aria-label="anchor" href="#bivariate"></a>
</h3>
<p>A list of all bivariate global methods can be seen here:</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/listSFEMethods.html" class="external-link">listSFEMethods</a></span><span class="op">(</span>variate <span class="op">=</span> <span class="st">"bi"</span>, scope <span class="op">=</span> <span class="st">"global"</span><span class="op">)</span></span>
<span><span class="co">#&gt;                  name                                     description</span></span>
<span><span class="co">#&gt; 1                 lee                       Lee's bivariate statistic</span></span>
<span><span class="co">#&gt; 2              lee.mc Lee's bivariate static with permutation testing</span></span>
<span><span class="co">#&gt; 3            lee.test                                    Lee's L test</span></span>
<span><span class="co">#&gt; 4     cross_variogram                                 Cross variogram</span></span>
<span><span class="co">#&gt; 5 cross_variogram_map                             Cross variogram map</span></span></code></pre></div>
<p>When calling <code>calculate*variate()</code> or
<code>run*variate()</code>, the <code>type</code> (2nd) argument takes
either an <code>SFEMethod</code> object or a string that matches an
entry in the <code>name</code> column in the data frame returned by
<code><a href="https://rdrr.io/pkg/Voyager/man/listSFEMethods.html" class="external-link">listSFEMethods()</a></code>.</p>
<div class="section level4">
<h4 id="lees-l">Lee’s L<a class="anchor" aria-label="anchor" href="#lees-l"></a>
</h4>
<p>Lee’s L <span class="citation">(Lee 2001)</span> was developed from
relating Moran’s I to Pearson correlation, and is defined as</p>
<p><span class="math display">\[
L_{X,Y} = \frac{n}{\sum_{i=1}^n \sum_{j=1}^n w_{ij}} \frac{\sum_{i=1}^n
\left[ \sum_{j=1}^n w_{ij}  (x_j - \bar{x}) \right] \left[ \sum_{j=1}^n
w_{ij} (y_j - \bar{y}) \right]}{\sqrt{\sum_{i=1}^n (x_i -
\bar{x})^2}\sqrt{\sum_{i=1}^n (y_i - \bar{y})^2} },
\]</span></p>
<p>where <span class="math inline">\(n\)</span> is the number of spots
or locations, <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span> are different locations, or spots in
the Visium context, <span class="math inline">\(x\)</span> and <span class="math inline">\(y\)</span> are variables with values at each
location, and <span class="math inline">\(w_{ij}\)</span> is a spatial
weight, which can be inversely proportional to distance between spots or
an indicator of whether two spots are neighbors, subject to various
definitions of neighborhood.</p>
<p>Here we compute Lee’s L for top highly variagle genes (HVGs) in this
dataset:</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hvgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, fdr.threshold <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span></span></code></pre></div>
<p>Because bivariate global results can have very different formats
(matrix for Lee’s L and lists for many other methods), the results are
not stored in the SFE object.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/calculateBivariate.html" class="external-link">calculateBivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, type <span class="op">=</span> <span class="st">"lee"</span>, feature1 <span class="op">=</span> <span class="va">hvgs</span><span class="op">)</span></span></code></pre></div>
<p>This gives a spatially informed correlation matrix among the genes,
which can be plotted as a heatmap:</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pal_rng</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/getDivergeRange.html" class="external-link">getDivergeRange</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span>
<span><span class="va">pal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scico/man/scico.html" class="external-link">scico</a></span><span class="op">(</span><span class="fl">256</span>, begin <span class="op">=</span> <span class="va">pal_rng</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, end <span class="op">=</span> <span class="va">pal_rng</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, palette <span class="op">=</span> <span class="st">"vik"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span><span class="va">res</span>, color <span class="op">=</span> <span class="va">pal</span>, show_rownames <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>         show_colnames <span class="op">=</span> <span class="cn">FALSE</span>, cellwidth <span class="op">=</span> <span class="fl">1</span>, cellheight <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-53-1.png" width="576"></p>
<p>Some coexpression blocks can be seen. Note that unlike in Pearson
correlation, the diagonal is not 1, because</p>
<p><span class="math display">\[
L_{X,X} = \frac{\sum_i (\tilde x_i - \bar x)^2}{\sum_i (x_i - \bar x)^2}
= \mathrm{SSS}_X,
\]</span></p>
<p>which is approximated the ratio between the variance of spatially
lagged <span class="math inline">\(x\)</span> and variance of <span class="math inline">\(x\)</span>. Because the spatial lag introduces
smoothing, the spatial lag reduced variance, making the diagonal less
than 1. This is the spatial smoothing scalar (SSS), and Moran’s I is
approximately Pearson correlation between <span class="math inline">\(X\)</span> and spatially lagged <span class="math inline">\(X\)</span> (<span class="math inline">\(\tilde
X\)</span>) multiplied by SSS:</p>
<p><span class="math display">\[
I = \mathrm{SSS}_X \cdot \rho_{X, \tilde X}
\]</span></p>
<p>Similarly for Lee’s L, as shown in <span class="citation">(Lee
2001)</span>,</p>
<p><span class="math display">\[
L_{X, Y} = \sqrt{\mathrm{SSS}_X}\sqrt{\mathrm{SSS}_Y} \cdot \rho_{\tilde
X, \tilde Y}
\]</span></p>
<p>With more spatial clustering, the variance is less reduced by the
spatial lag, leading to a larger SSS.</p>
<p>Weighted correlation network analysis (WGCNA) <span class="citation">(Langfelder and Horvath 2008)</span> is a time honored
method to find gene co-expression modules, and it can take any
correlation matrix. Then it would be interesting to apply WGCNA to the
Lee’s L matrix to identify spatially informed gene co-expression
modules.</p>
</div>
</div>
<div class="section level3">
<h3 id="multivariate">Multivariate<a class="anchor" aria-label="anchor" href="#multivariate"></a>
</h3>
<div class="section level4">
<h4 id="non-spatial-pca">Non-spatial PCA<a class="anchor" aria-label="anchor" href="#non-spatial-pca"></a>
</h4>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hvgs2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">dec</span>, n <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span></code></pre></div>
<p>First we run non-spatial PCA, to compare to MULTISPATI.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">29</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span></span>
<span>    <span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, ncomponents <span class="op">=</span> <span class="fl">20</span>, subset_row <span class="op">=</span> <span class="va">hvgs2</span>,</span>
<span>                         exprs_values <span class="op">=</span> <span class="st">"logcounts"</span>, scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt;    user  system elapsed </span></span>
<span><span class="co">#&gt;   0.395   0.068   0.274</span></span></code></pre></div>
<p>That’s pretty quick for almost 400,000 cells, but there aren’t that
many genes here. Use the elbow plot to see variance explained by each
PC:</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/ElbowPlot.html" class="external-link">ElbowPlot</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-56-1.png" width="700"></p>
<p>Plot top gene loadings in each PC</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotDimLoadings.html" class="external-link">plotDimLoadings</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, swap_rownames <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-57-1.png" width="700"></p>
<p>Many of these genes seem to be related to the endothelium. PC1 and
PC4 concern the Kupffer cells as well, as the Kupffer cell marker gene
Cdh5 has high loading.</p>
<p>Plot the first 4 PCs in space</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/spatialReducedDim.html" class="external-link">spatialReducedDim</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"PCA"</span>, <span class="fl">4</span>, </span>
<span>                  divergent <span class="op">=</span> <span class="cn">TRUE</span>, diverge_center <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-58-1.png" width="768"></p>
</div>
<div class="section level4">
<h4 id="multispati-pca">MULTISPATI PCA<a class="anchor" aria-label="anchor" href="#multispati-pca"></a>
</h4>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/calculateMultivariate.html" class="external-link">runMultivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"multispati"</span>, colGraphName <span class="op">=</span> <span class="st">"visium"</span>,</span>
<span>                                  nfposi <span class="op">=</span> <span class="fl">20</span>, nfnega <span class="op">=</span> <span class="fl">20</span>, subset_row <span class="op">=</span> <span class="va">hvgs2</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: useNames = NA is deprecated. Instead, specify either useNames = TRUE</span></span>
<span><span class="co">#&gt; or useNames = TRUE.</span></span>
<span><span class="co">#&gt;    user  system elapsed </span></span>
<span><span class="co">#&gt;   0.582   0.100   0.426</span></span></code></pre></div>
<p>Then plot the most positive and most negative eigenvalues. Note that
the eigenvalues here are not variance explained. Instead, they are the
product of variance explained and Moran’s I. So the most positive
eigenvalues correspond to eigenvectors that simultaneously explain more
variance and have large positive Moran’s I. The most negative
eigenvalues correspond to eigenvectors that simultaneously explain more
variance and have negative Moran’s I.</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/ElbowPlot.html" class="external-link">ElbowPlot</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, nfnega <span class="op">=</span> <span class="fl">20</span>, reduction <span class="op">=</span> <span class="st">"multispati"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-60-1.png" width="700"></p>
<p>Here the positive eigenvalues drop sharply from PC1 to PC4, and there
is only one very negative eigenvalue which might be interesting, which
is unsurprising given the moderately negative Moran’s I for
<code>nCounts</code> and <code>nGenes</code>. However, from the <a href="https://pachterlab.github.io/voyager/articles/vig6_merfish.html#morans-i" class="external-link">first
MERFISH vignette</a>, none of the genes have very negative Moran’s I.
Perhaps the negative eigenvalue comes from negative spatial
autocorrelation in a gene program or “eigengene” and is not obvious from
individual genes. This is the beauty of multivariate analysis.</p>
<p>What do these components mean? Each component is a linear combination
of genes to maximize the product of variance explained and Moran’s I.
The second component maximizes this product provided that it’s
orthogonal to the first component, and so on. As the loss in variance
explained is usually not huge, these components can be considered axes
along which <em>spatially coherent</em> groups of spots are separated
from each other as much as possible according to expression of the
highly variable genes, so in theory, clustering with positive MULTISPATI
components should give more spatially coherent clusters. Because of the
spatial coherence, MULTISPATI might be more robust to outliers.</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotDimLoadings.html" class="external-link">plotDimLoadings</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, reduction <span class="op">=</span> <span class="st">"multispati"</span>,</span>
<span>                swap_rownames <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-61-1.png" width="700"></p>
<p>Plot the these PCs:</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/spatialReducedDim.html" class="external-link">spatialReducedDim</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"multispati"</span>, <span class="fl">4</span>, </span>
<span>                  divergent <span class="op">=</span> <span class="cn">TRUE</span>, diverge_center <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-62-1.png" width="768"></p>
</div>
<div class="section level4">
<h4 id="spatial-autocorrelation-of-principal-components">Spatial autocorrelation of principal components<a class="anchor" aria-label="anchor" href="#spatial-autocorrelation-of-principal-components"></a>
</h4>
<p>Here we compare Moran’s I for cell embeddings in each non-spatial and
MULTISPATI PC:</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># non-spatial</span></span>
<span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">reducedDimMoransI</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, dimred <span class="op">=</span> <span class="st">"PCA"</span>, components <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span></span>
<span><span class="co"># spatial</span></span>
<span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">reducedDimMoransI</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, dimred <span class="op">=</span> <span class="st">"multispati"</span>, components <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">40</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df_moran</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>PCA <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/colFeatureData.html" class="external-link">reducedDimFeatureData</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"PCA"</span><span class="op">)</span><span class="op">$</span><span class="va">moran_Vis5A</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">]</span>,</span>
<span>                   MULTISPATI_pos <span class="op">=</span> </span>
<span>                       <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/colFeatureData.html" class="external-link">reducedDimFeatureData</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"multispati"</span><span class="op">)</span><span class="op">$</span><span class="va">moran_Vis5A</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">]</span>,</span>
<span>                   MULTISPATI_neg <span class="op">=</span> </span>
<span>                       <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/colFeatureData.html" class="external-link">reducedDimFeatureData</a></span><span class="op">(</span><span class="va">sfe_tissue</span>,<span class="st">"multispati"</span><span class="op">)</span><span class="op">$</span><span class="va">moran_Vis5A</span><span class="op">[</span><span class="fl">21</span><span class="op">:</span><span class="fl">40</span><span class="op">]</span> <span class="op">|&gt;</span> </span>
<span>                       <span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>                   index <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"ditto_colors"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">mb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/moranBounds.html" class="external-link">moranBounds</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"visium"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;       Imin       Imax </span></span>
<span><span class="co">#&gt; -0.5762132  1.0021884</span></span></code></pre></div>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df_moran</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="va">index</span>, values_to <span class="op">=</span> <span class="st">"value"</span>, names_to <span class="op">=</span> <span class="st">"name"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">index</span>, <span class="va">value</span>, color <span class="op">=</span> <span class="va">name</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">ditto_colors</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">0</span>, color <span class="op">=</span> <span class="st">"gray"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="va">mb</span>, linetype <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="st">"gray"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/breaks_pretty.html" class="external-link">breaks_pretty</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/breaks_width.html" class="external-link">breaks_width</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"Moran's I"</span>, color <span class="op">=</span> <span class="st">"Type"</span>, x <span class="op">=</span> <span class="st">"Component"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workshop_files/figure-html/unnamed-chunk-67-1.png" width="700"></p>
<p>In MULTISPATI, Moran’s I is high in PC1 and PC2, but then sharply
drops. Moran’s I for the PC with the most negative eigenvalues is not
very negative, which means the large magnitude of that eigenvalue comes
from explaining more variance. However, considering the lower bound of
Moran’s I that is around -0.6 instead of -1, the magnitude of Moran’s I
for the PC with the most negative eigenvalue is not trivial.</p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">df_moran</span><span class="op">$</span><span class="va">MULTISPATI_neg</span><span class="op">)</span> <span class="op">/</span> <span class="va">mb</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="co">#&gt;      Imin </span></span>
<span><span class="co">#&gt; 0.7553984</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="caveats">Caveats<a class="anchor" aria-label="anchor" href="#caveats"></a>
</h3>
<ol style="list-style-type: decimal">
<li>In the current version of <code>Voyager</code>, only univariate
spatial autocorrelation metrics are supported. Anisotropy, bivariate,
and multivariate spatial analyses will be added in later versions.</li>
<li>The H&amp;E image can alter perception of the colors of the
geometries.</li>
<li>Only 2D data is supported at present, although in principle,
<code>sf</code> and <code>GEOS</code> support 3D data.</li>
<li>Spatial neighborhoods only make sense within the same tissue
section. Then what to do with multiple tissue sections, from biological
replica, and from different conditions? For the mouse brain, different
biological replica can be registered to the Allen Common Coordinate
Framework (CCF) to be spatially comparable. Indeed, it would be
interesting to see the biological variability of healthy wild type gene
expression at the same fine scaled region in the brain. However, there
is no CCF for tissues without a stereotypical structure, such as adipose
and skeletal muscle. We don’t have a good solution to spatially compare
different tissue sections yet. Perhaps global spatial statistics over
the whole section or histological regions within the section can be
compared. The problem remains to select the most informative metrics to
compare. Perhaps a spatially-informed dimension reduction method, taking
not only the gene count matrix, but also the adjacency matrices of the
spatial neighborhood graphs (different sections will be different blocks
in the matrix) projecting the cells or Visium spots from different
sections into a shared low dimensional space can facilitate the
comparison. Here batch effect must be corrected, and the dimension
reduction should be interpretable, and scalable.</li>
</ol>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Etc/UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] pheatmap_1.0.12                scico_1.4.0                   </span></span>
<span><span class="co">#&gt;  [3] tidyr_1.3.0                    tibble_3.2.1                  </span></span>
<span><span class="co">#&gt;  [5] BiocParallel_1.35.3            patchwork_1.1.2               </span></span>
<span><span class="co">#&gt;  [7] scales_1.2.1                   sf_1.0-14                     </span></span>
<span><span class="co">#&gt;  [9] Matrix_1.6-0                   rjson_0.2.21                  </span></span>
<span><span class="co">#&gt; [11] scater_1.29.0                  ggplot2_3.4.2                 </span></span>
<span><span class="co">#&gt; [13] scran_1.29.0                   scuttle_1.11.0                </span></span>
<span><span class="co">#&gt; [15] SFEData_1.3.0                  Voyager_1.3.0                 </span></span>
<span><span class="co">#&gt; [17] SingleCellExperiment_1.23.0    SummarizedExperiment_1.31.1   </span></span>
<span><span class="co">#&gt; [19] Biobase_2.61.0                 GenomicRanges_1.53.1          </span></span>
<span><span class="co">#&gt; [21] GenomeInfoDb_1.37.2            IRanges_2.35.2                </span></span>
<span><span class="co">#&gt; [23] S4Vectors_0.39.1               BiocGenerics_0.47.0           </span></span>
<span><span class="co">#&gt; [25] MatrixGenerics_1.13.0          matrixStats_1.0.0             </span></span>
<span><span class="co">#&gt; [27] SpatialFeatureExperiment_1.3.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] later_1.3.1                   bitops_1.0-7                 </span></span>
<span><span class="co">#&gt;   [3] filelock_1.0.2                R.oo_1.25.0                  </span></span>
<span><span class="co">#&gt;   [5] lifecycle_1.0.3               edgeR_3.43.7                 </span></span>
<span><span class="co">#&gt;   [7] rprojroot_2.0.3               lattice_0.21-8               </span></span>
<span><span class="co">#&gt;   [9] magrittr_2.0.3                limma_3.57.6                 </span></span>
<span><span class="co">#&gt;  [11] sass_0.4.7                    rmarkdown_2.23               </span></span>
<span><span class="co">#&gt;  [13] jquerylib_0.1.4               yaml_2.3.7                   </span></span>
<span><span class="co">#&gt;  [15] metapod_1.9.0                 httpuv_1.6.11                </span></span>
<span><span class="co">#&gt;  [17] sp_2.0-0                      RColorBrewer_1.1-3           </span></span>
<span><span class="co">#&gt;  [19] DBI_1.1.3                     zlibbioc_1.47.0              </span></span>
<span><span class="co">#&gt;  [21] purrr_1.0.1                   R.utils_2.12.2               </span></span>
<span><span class="co">#&gt;  [23] RCurl_1.98-1.12               rappdirs_0.3.3               </span></span>
<span><span class="co">#&gt;  [25] GenomeInfoDbData_1.2.10       ggrepel_0.9.3                </span></span>
<span><span class="co">#&gt;  [27] irlba_2.3.5.1                 terra_1.7-39                 </span></span>
<span><span class="co">#&gt;  [29] units_0.8-2                   RSpectra_0.16-1              </span></span>
<span><span class="co">#&gt;  [31] dqrng_0.3.0                   pkgdown_2.0.7                </span></span>
<span><span class="co">#&gt;  [33] DelayedMatrixStats_1.23.0     codetools_0.2-19             </span></span>
<span><span class="co">#&gt;  [35] DropletUtils_1.21.0           DelayedArray_0.27.9          </span></span>
<span><span class="co">#&gt;  [37] tidyselect_1.2.0              farver_2.1.1                 </span></span>
<span><span class="co">#&gt;  [39] ScaledMatrix_1.9.1            viridis_0.6.3                </span></span>
<span><span class="co">#&gt;  [41] BiocFileCache_2.9.1           jsonlite_1.8.7               </span></span>
<span><span class="co">#&gt;  [43] BiocNeighbors_1.19.0          e1071_1.7-13                 </span></span>
<span><span class="co">#&gt;  [45] ellipsis_0.3.2                systemfonts_1.0.4            </span></span>
<span><span class="co">#&gt;  [47] tools_4.3.1                   ggnewscale_0.4.9             </span></span>
<span><span class="co">#&gt;  [49] ragg_1.2.5                    Rcpp_1.0.11                  </span></span>
<span><span class="co">#&gt;  [51] glue_1.6.2                    gridExtra_2.3                </span></span>
<span><span class="co">#&gt;  [53] SparseArray_1.1.10            xfun_0.39                    </span></span>
<span><span class="co">#&gt;  [55] dplyr_1.1.2                   HDF5Array_1.29.3             </span></span>
<span><span class="co">#&gt;  [57] withr_2.5.0                   BiocManager_1.30.21.1        </span></span>
<span><span class="co">#&gt;  [59] fastmap_1.1.1                 boot_1.3-28.1                </span></span>
<span><span class="co">#&gt;  [61] rhdf5filters_1.13.4           bluster_1.11.3               </span></span>
<span><span class="co">#&gt;  [63] fansi_1.0.4                   spData_2.3.0                 </span></span>
<span><span class="co">#&gt;  [65] digest_0.6.33                 rsvd_1.0.5                   </span></span>
<span><span class="co">#&gt;  [67] R6_2.5.1                      mime_0.12                    </span></span>
<span><span class="co">#&gt;  [69] textshaping_0.3.6             colorspace_2.1-0             </span></span>
<span><span class="co">#&gt;  [71] wk_0.7.3                      RSQLite_2.3.1                </span></span>
<span><span class="co">#&gt;  [73] R.methodsS3_1.8.2             utf8_1.2.3                   </span></span>
<span><span class="co">#&gt;  [75] generics_0.1.3                class_7.3-22                 </span></span>
<span><span class="co">#&gt;  [77] httr_1.4.6                    S4Arrays_1.1.4               </span></span>
<span><span class="co">#&gt;  [79] spdep_1.2-8                   pkgconfig_2.0.3              </span></span>
<span><span class="co">#&gt;  [81] gtable_0.3.3                  blob_1.2.4                   </span></span>
<span><span class="co">#&gt;  [83] XVector_0.41.1                htmltools_0.5.5              </span></span>
<span><span class="co">#&gt;  [85] png_0.1-8                     SpatialExperiment_1.11.0     </span></span>
<span><span class="co">#&gt;  [87] knitr_1.43                    curl_5.0.1                   </span></span>
<span><span class="co">#&gt;  [89] proxy_0.4-27                  cachem_1.0.8                 </span></span>
<span><span class="co">#&gt;  [91] rhdf5_2.45.1                  stringr_1.5.0                </span></span>
<span><span class="co">#&gt;  [93] BiocVersion_3.18.0            KernSmooth_2.23-22           </span></span>
<span><span class="co">#&gt;  [95] parallel_4.3.1                vipor_0.4.5                  </span></span>
<span><span class="co">#&gt;  [97] AnnotationDbi_1.63.2          desc_1.4.2                   </span></span>
<span><span class="co">#&gt;  [99] s2_1.1.4                      pillar_1.9.0                 </span></span>
<span><span class="co">#&gt; [101] grid_4.3.1                    vctrs_0.6.3                  </span></span>
<span><span class="co">#&gt; [103] promises_1.2.0.1              BiocSingular_1.17.1          </span></span>
<span><span class="co">#&gt; [105] dbplyr_2.3.3                  beachmat_2.17.13             </span></span>
<span><span class="co">#&gt; [107] xtable_1.8-4                  cluster_2.1.4                </span></span>
<span><span class="co">#&gt; [109] beeswarm_0.4.0                evaluate_0.21                </span></span>
<span><span class="co">#&gt; [111] magick_2.7.4                  cli_3.6.1                    </span></span>
<span><span class="co">#&gt; [113] locfit_1.5-9.8                compiler_4.3.1               </span></span>
<span><span class="co">#&gt; [115] rlang_1.1.1                   crayon_1.5.2                 </span></span>
<span><span class="co">#&gt; [117] labeling_0.4.2                classInt_0.4-9               </span></span>
<span><span class="co">#&gt; [119] fs_1.6.2                      ggbeeswarm_0.7.2             </span></span>
<span><span class="co">#&gt; [121] stringi_1.7.12                viridisLite_0.4.2            </span></span>
<span><span class="co">#&gt; [123] deldir_1.0-9                  munsell_0.5.0                </span></span>
<span><span class="co">#&gt; [125] Biostrings_2.69.2             ExperimentHub_2.9.1          </span></span>
<span><span class="co">#&gt; [127] sparseMatrixStats_1.13.0      bit64_4.0.5                  </span></span>
<span><span class="co">#&gt; [129] Rhdf5lib_1.23.0               KEGGREST_1.41.0              </span></span>
<span><span class="co">#&gt; [131] statmod_1.5.0                 shiny_1.7.4.1                </span></span>
<span><span class="co">#&gt; [133] highr_0.10                    interactiveDisplayBase_1.39.0</span></span>
<span><span class="co">#&gt; [135] AnnotationHub_3.9.1           igraph_1.5.0                 </span></span>
<span><span class="co">#&gt; [137] memoise_2.0.1                 bslib_0.5.0                  </span></span>
<span><span class="co">#&gt; [139] bit_4.0.5</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Anselin1995-cs" class="csl-entry">
Anselin, Luc. 1995. <span>“Local Indicators of Spatial
<span>Association—LISA</span>.”</span> <em>Geogr. Anal.</em> 27 (2):
93–115.
</div>
<div id="ref-Fotheringham2009-ak" class="csl-entry">
Fotheringham, A Stewart. 2009. <span>“<span>‘The Problem of Spatial
Autocorrelation’</span> and Local Spatial Statistics.”</span> <em>Geogr.
Anal.</em> 41 (4): 398–403.
</div>
<div id="ref-Langfelder2008-fs" class="csl-entry">
Langfelder, Peter, and Steve Horvath. 2008. <span>“<span>WGCNA</span>:
An <span>R</span> Package for Weighted Correlation Network
Analysis.”</span> <em>BMC Bioinformatics</em> 9 (December): 559.
</div>
<div id="ref-Lee2001-tm" class="csl-entry">
Lee, Sang-Il. 2001. <span>“Developing a Bivariate Spatial Association
Measure: An Integration of Pearson’s r and Moran’s
<span>I</span>.”</span> <em>J. Geogr. Syst.</em> 3 (4): 369–85.
</div>
<div id="ref-Lun2016-yq" class="csl-entry">
Lun, Aaron T L, Karsten Bach, and John C Marioni. 2016. <span>“Pooling
Across Cells to Normalize Single-Cell <span>RNA</span> Sequencing Data
with Many Zero Counts.”</span> <em>Genome Biol.</em> 17 (April): 75.
</div>
<div id="ref-McKellar2021-ek" class="csl-entry">
McKellar, David W, Lauren D Walter, Leo T Song, Madhav Mantri, Michael F
Z Wang, Iwijn De Vlaminck, and Benjamin D Cosgrove. 2021.
<span>“Large-Scale Integration of Single-Cell Transcriptomic Data
Captures Transitional Progenitor States in Mouse Skeletal Muscle
Regeneration.”</span> <em>Commun Biol</em> 4 (1): 1280.
</div>
<div id="ref-Wang2017-li" class="csl-entry">
Wang, Chao, Feng Yue, and Shihuan Kuang. 2017. <span>“Muscle Histology
Characterization Using <span>H&amp;E</span> Staining and Muscle Fiber
Type Classification Using Immunofluorescence Staining.”</span> <em>Bio
Protoc</em> 7 (10).
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Lambda Moses.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
